src/main/java/mezz/jei/startup/ClientLifecycleHandler.java

```diff
@@ -89,43 +90,63 @@ public ClientLifecycleHandler(NetworkHandler networkHandler, Textures textures)
 		KeyBindings.init();
 
 		EventBusHelper.addListener(this, WorldEvent.Save.class, event -> worldConfig.onWorldSave());
+
+		EventBusHelper.addListener(this, AddReloadListenerEvent.class, this::reloadListenerSetup);
+		// Listener for integrated server
 		EventBusHelper.addListener(this, RecipesUpdatedEvent.class, event -> {
+			if (Minecraft.getInstance().isIntegratedServerRunning()) {
+				setupJEI();
+			}
+		});
+		// Listener for vanilla remote server
+		EventBusHelper.addListener(this, TagsUpdatedEvent.VanillaTagTypes.class, event -> {
+			ClientPlayNetHandler connection = Minecraft.getInstance().getConnection();
+			if (connection != null && NetworkHooks.isVanillaConnection(connection.getNetworkManager())) {
+				setupJEI();
+			}
+		});
+		// Listener for forge remote server
+		EventBusHelper.addListener(this, TagsUpdatedEvent.CustomTagTypes.class, event -> {
 			ClientPlayNetHandler connection = Minecraft.getInstance().getConnection();
-			if (connection != null) {
-				NetworkManager networkManager = connection.getNetworkManager();
-				worldConfig.syncWorldConfig(networkManager);
+			if (connection != null && !NetworkHooks.isVanillaConnection(connection.getNetworkManager()) && !Minecraft.getInstance().isIntegratedServerRunning()) {
```

Comment:
To help clarify this logic, can you create a new enum and a helper function that returns `IntegratedVanillaServer`, `IntegratedForgeServer`, `RemoteVanillaServer`, or `RemoteForgeServer`? Then these conditions can just check using the enums.